controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -0,0 +1,745 @@
+            Object api = reg.getCapability().getRuntimeAPI();
+            if (api == null) {
+                throw ControllerLogger.MGMT_OP_LOGGER.capabilityDoesNotExposeRuntimeAPI(capabilityName);
+            }
+            return apiType.cast(api);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<CapabilityRegistration> getCapabilities() {
+        try {
+            readLock.lock();
+            return Collections.unmodifiableSet(new HashSet<>(capabilities.values()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<CapabilityRegistration> getPossibleCapabilities() {
+        try {
+            readLock.lock();
+            return Collections.unmodifiableSet(new HashSet<>(possibleCapabilities.values()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public ServiceName getCapabilityServiceName(String capabilityName, CapabilityContext context, Class<?> serviceType) {
+        // Here we can't know the dependent name. So this can only be called when resolution is complete.
+        assert resolutionContext.resolutionComplete;
+        try {
+            readLock.lock();
+            RuntimeCapabilityRegistration reg = getCapabilityRegistration(capabilityName, context);
+            RuntimeCapability<?> cap = reg.getCapability();
+            return cap.getCapabilityServiceName(serviceType);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    //end ImmutableCapabilityRegistry methods
+
+    /**
+     * Publish the changes to main registry
+     */
+    void publish() {
+        assert publishedFullRegistry != null : "Cannot write directly to main registry";
+
+        if (!modified) {
+            return;
+        }
+        try {
+            writeLock.lock();
+            CapabilityRegistry published = publishedFullRegistry;
+            published.capabilities.clear();
+            published.capabilities.putAll(capabilities);
+
+            published.possibleCapabilities.clear();
+            published.possibleCapabilities.putAll(possibleCapabilities);
+
+            published.requirements.clear();
+            published.requirements.putAll(requirements);
+
+            published.runtimeOnlyRequirements.clear();
+            published.runtimeOnlyRequirements.putAll(runtimeOnlyRequirements);
+            modified = false;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Discard the changes.
+     */
+    void rollback() {
+        if (publishedFullRegistry == null) {
+            return;
+        }
+        try {
+            writeLock.lock();
+            clear();
+            copy(publishedFullRegistry, this);
+            modified = false;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    boolean isModified() {
+        return modified;
```

Comment:
Does it really? modifed field only gets changed on write operations and should affect read ops.